File: cloudflareapi.py
# -*- coding: utf-8 -*-
# cloudflareapi.py
# Copyright (C) 2020 KunoiSayami and contributors
#
# This module is part of passive-DDNS and is released under
# the AGPL v3 License: https://www.gnu.org/licenses/agpl-3.0.txt
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.
import ast
import logging
import time
from configparser import ConfigParser
from dataclasses import dataclass
from typing import Dict, Generator, List, TypeVar

import requests

FixedType = TypeVar('FixedType', bool, str, int)

logger: logging.Logger = logging.getLogger('CloudFlareApi')
logger.setLevel(logging.INFO)
retry_delay = [60, 20, 10, 0]

@dataclass
class DNSRecord:
	id: str
	zone_id: str
	name: str
	content: str
	proxied: bool
	ttl: int

	@classmethod
	def create(cls, rt: Dict[str, FixedType]) -> 'DNSRecord':
		self = cls(rt['id'], rt['zone_id'], rt['name'], rt['content'], rt['proxied'], rt['ttl'])
		return self

	def update_dns(self, session: requests.Session, change_to_ip: str, retries: int=3) -> None:
		self.content = change_to_ip
		time.sleep(retry_delay[retries])
		try:
			r = session.put(f'https://api.cloudflare.com/client/v4/zones/{self.zone_id}/dns_records/{self.id}', json=self.get_params(), timeout=30)
			r.raise_for_status()
			r.close()
		except requests.exceptions.Timeout:
			if retries > 0:
				logger.error('Got Timeout error during update dns record (retries: %d)', retries)
				self.update_dns(session, change_to_ip, retries - 1)
			else:
				logger.critical('Can\'t update dns, raise it')
				raise

	def get_params(self) -> Dict[str, FixedType]:
		return dict(type='A', name=self.name, content=self.content, proxied=self.proxied, ttl=self.ttl)


class CloudFlareApi:
	def __init__(self, config: ConfigParser):
		self.domains: Dict[str, List[str]]
		self.api_token: str = config.get('cloudflare', 'token')
		if config.has_option('cloudflare', 'header_domain'):
			self.domains = ast.literal_eval(config.get('cloudflare', 'header_domain'))
		else:
			self.domains = ast.literal_eval(config.get('cloudflare', 'ipv4_domain'))
		self.session = requests.Session()
		self.session.headers.update({
			'Authorization': f'Bearer {self.api_token}',
			'Content-Type': 'application/json',
			'Connection': 'close'
			})

	def get_domain_record(self, domain: str, name: str, retries: int=3) -> Dict[str, FixedType]:
		time.sleep(retry_delay[retries])
		try:
			r = self.session.get(f'https://api.cloudflare.com/client/v4/zones/{domain}/dns_records', params={'type': 'A', 'name': name}, timeout=30)
			r.raise_for_status()
			result = r.json()['result'][0]
			r.close()
			return result
		except requests.exceptions.Timeout:
			if retries > 0:
				logger.error('Got timeout error (retries: %d)', retries)
				return self.get_domain_record(domain, name, retries - 1)
			else:
				logger.critical('Can\'t fetch dns records, raise it')
				raise

	def get_records(self) -> Generator[DNSRecord, None, None]:
		for key, item in self.domains.items():
			for x in item:
				yield DNSRecord.create(self.get_domain_record(key, x))

	def update_records(self, change_to_ip: str) -> bool:
		is_update: bool = False
		for record in self.get_records():
			if record.content != change_to_ip:
				is_update = True
				record.update_dns(self.session, change_to_ip)
		return is_update

	def close(self) -> None:
		self.session.close()


def test():
	config = ConfigParser()
	config.read('data/config.ini')
	c = CloudFlareApi(config)
	for record in c.get_records():
		print(record)
	c.close()

if __name__ == "__main__":
	test()

File: rust/src/cloudflare_api.rs
/*
 ** Copyright (C) 2021 KunoiSayami
 **
 ** This file is part of passive-DDNS and is released under
 ** the AGPL v3 License: https://www.gnu.org/licenses/agpl-3.0.txt
 **
 ** This program is free software: you can redistribute it and/or modify
 ** it under the terms of the GNU Affero General Public License as published by
 ** the Free Software Foundation, either version 3 of the License, or
 ** any later version.
 **
 ** This program is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 ** GNU Affero General Public License for more details.
 **
 ** You should have received a copy of the GNU Affero General Public License
 ** along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
pub(crate) mod api {
    use std::collections::HashMap;
    use serde_derive::{Deserialize, Serialize};

    #[derive(Deserialize)]
    struct DNSRecord {
        id: String,
        zone_id: String,
        name: String,
        content: String,
        proxied: bool,
        ttl: i32,
    }

    impl DNSRecord {
        fn update_content<T>(&mut self, content: T)
            where T: Into<String> {
            self.content = content.into();
        }

        fn update_ns_record(&self, session: &reqwest::blocking::Client) -> bool {
            let resp = session.put(
                format!("https://api.cloudflare.com/client/v4/zones/{}/dns_records/{}",
                        &self.zone_id,
                        &self.id
                ).as_str()).json(&PutDNSRecord::from(self))
                .send()
                .unwrap();
            resp.status().is_success()
        }
    }

    #[derive(Serialize)]
    struct PutDNSRecord {
        #[serde(rename="type")]
        t: String,
        name: String,
        content: String,
        proxied: bool,
        ttl: i32,
    }

    impl PutDNSRecord {
        fn from(dns_record: &DNSRecord) -> PutDNSRecord {
            PutDNSRecord{
                t: String::from('A'),
                name: String::from(&dns_record.name),
                content: String::from(&dns_record.content),
                proxied: dns_record.proxied,
                ttl: dns_record.ttl
            }
        }
    }

    struct Zone {
        zone_id: String,
        domains: Vec<String>
    }

    impl Zone {
        pub fn new<T>(original_string: T) -> Zone
            where T: Into<String> {
            let basic_re = regex::Regex::new(r"'([a-f\d]+)':\s*\[(('[\w\.]+',\s*)*'[\w\.]+')\]").unwrap();
            let domain_re = regex::Regex::new(r"([\w\.]+)").unwrap();
            let original_string = original_string.into();
            log::debug!("Parse string: {}", &original_string);
            let cap = basic_re.captures(original_string.as_str()).unwrap();
            let zone_id = String::from(&cap[1]);
            log::debug!("Processing zone: {}", zone_id);
            let mut domains: Vec<String> = vec![];
            for cap in domain_re.captures_iter(&cap[2]) {
                let domain = String::from(&cap[1]);
                domains.push(domain.clone());
                log::debug!("Push {} to {}", domain, zone_id);
            }
            Zone{zone_id, domains}
        }

        pub fn request_domain_record(&self, session: &reqwest::blocking::Client) -> Vec<DNSRecord>{
            let mut records: Vec<DNSRecord> = Default::default();
            //let form: HashMap::<_, _>::from_iter = (("test", "test"), ("test", "test"));

            for domain in &self.domains {
                let name = domain.as_str();
                let query: HashMap<&str, &str> = [("type", "A"), ("name", name)].iter().cloned().collect();
                let resp = session.get(format!("https://api.cloudflare.com/client/v4/zones/{}/dns_records", &self.zone_id).as_str())
                    .query(&query)
                    .send()
                    .unwrap();
                let resp_json: serde_json::Value = resp.json().unwrap();
                let dns_record: DNSRecord = serde_json::from_value(resp_json["result"][0].clone())
                    .unwrap();
                records.push(dns_record);
            }
            records
        }
    }

    pub(crate) struct Configure {
        zones: Vec<Zone>,
        api_token: String,
        session: reqwest::blocking::Client
    }

    impl Configure {
        pub fn new<T>(domains: T, api_token: T) -> Configure
            where T: Into<String> {
            let re = regex::Regex::new(r"('[a-f\d]+':\s*\[('[\w\.]+',\s*)*'[\w\.]+'\])").unwrap();
            let original_domain_string = domains.into();
            let mut zones: Vec<Zone> = vec![];
            for cap in re.captures_iter(&original_domain_string.as_str()) {
                let domain_configure = String::from(&cap[1]);
                zones.push(Zone::new(domain_configure));
            }
            let api_token = api_token.into();
            let mut header_map = reqwest::header::HeaderMap::new();
            header_map.insert("Authorization", format!("Bearer {}", api_token).parse().unwrap());
            header_map.insert("Content-Type", String::from("application/json").parse().unwrap());
            header_map.insert("Connection", String::from("close").parse().unwrap());

            let session = reqwest::blocking::Client::builder()
                .default_headers(header_map)
                .build()
                .unwrap();

            Configure{zones, api_token: api_token.into(), session}
        }

        fn fetch_data(&self) -> Vec<DNSRecord>{
            let mut result: Vec<DNSRecord> = Default::default();
            for zone in &self.zones {
                result.extend(zone.request_domain_record(&self.session));
            }
            result
        }

        pub fn update_dns_data(&self, new_data: String) {
            let mut need_updated: Vec<DNSRecord> = Default::default();
            for record in self.fetch_data() {
                if record.content != new_data {
                    let mut mut_record = record;
                    mut_record.content = String::from(&new_data);
                    need_updated.push(mut_record);
                }
            }
            for record in need_updated {
                record.update_ns_record(&self.session);
            }
        }
    }
}